**Zhafira Amanda - 24060121140100**

Dataset :Glass Identification

Link : https://archive.ics.uci.edu/dataset/42/glass+identification

# Pertemuan 2 - Evaluasi Algoritma

**1. Lakukan Eksplorasi terhadapat algoritma klasifikasi lain yang ada!**

In [1]:
import pandas as pd

In [2]:
url = "http://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"

In [3]:
names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'class']

In [4]:
dataset = pd.read_csv(url, names=names)

In [5]:
dataset.head(5)

sepal-length  sepal-width  petal-length  petal-width        class
0           5.1          3.5           1.4          0.2  Iris-setosa
1           4.9          3.0           1.4          0.2  Iris-setosa
2           4.7          3.2           1.3          0.2  Iris-setosa
3           4.6          3.1           1.5          0.2  Iris-setosa
4           5.0          3.6           1.4          0.2  Iris-setosa

**I. Validasi Dataset**

In [13]:
from sklearn import model_selection

In [14]:
# Split-out validation dataset
array = dataset.values
X = array[:,0:4]
Y = array[:,4]
validation_size = 0.20
seed = 7
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size,random_state=seed)

In [15]:
print("X_train:", X_train.shape)
print("X_validation:", X_validation.shape)
print("Y_train:", Y_train.shape)
print("Y_validation:", Y_validation.shape)

X_train: (120, 4)
X_validation: (30, 4)
Y_train: (120,)
Y_validation: (30,)


Pada tahap validasi dataset, dilakukan pembagian dataset menjadi dua subset: data pelatihan (training) dan data validasi. Dimana proses ini bertujuan untuk mengukur kinerja model pada data yang tidak digunakan selama pelatihan. Dapat dilihat diatas bahwa dataset awal dipecah menjadi X_train dan X_validation untuk fitur-fitur (atribut) serta Y_train dan Y_validation untuk label kelas. Setelah itu persentase atau bagian dari total validasi dataset diatur menggunakan parameter test_size yang disetel sebesar 20%, dengan nilai seed (random state) sebesar 7 untuk memastikan hasil. Hasil dari pembagian dataset ini adalah 120 sampel untuk pelatihan dan 30 sampel untuk validasi, dengan masing-masing memiliki dimensi yang sesuai, yaitu 4 atribut untuk fitur dan 1 atribut untuk label kelas.

**II. K-Folds Cross Validation**

In [16]:
# Test options and evaluation metric
seed = 7
scoring = 'accuracy'

K-Folds Cross Validation adalah teknik pemodelan data yang membagi data menjadi beberapa fold, kemudian melatih model pada fold tertentu dan mengevaluasinya pada fold lainnya. Hal ini dilakukan berulang kali hingga semua fold telah digunakan untuk pelatihan dan evaluasi.Parameter **seed** digunakan untuk memastikan bahwa hasil cross validation tetap konsisten setiap kali dijalankan dan untuk parameter **scoring** digunakan untuk menentukan metrik evaluasi yang akan digunakan untuk mengukur performa model. Dalam kasus ini, metrik yang digunakan adalah akurasi. Dengan menggunakan K-Folds Cross Validation yang nantinya akan mendapatkan lebih akurat tentang performa model dibandingkan dengan hanya menggunakan data latih dan data uji.

**III. Membangun Model**

In [17]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import KFold, cross_val_score
from sklearn.preprocessing import StandardScaler

In [18]:
models = []

models.append(('DecisionTree', DecisionTreeClassifier()))
models.append(('RandomForest', RandomForestClassifier()))
models.append(('NeuralNetwork', MLPClassifier(max_iter=1000)))


In [19]:
# mengevaluasi setiap model secara bergantian
results = []
names = []

for name, model in models:
    kfold = KFold(n_splits=10, shuffle=True, random_state=seed)

    # Jika menggunakan Neural Network, lakukan penskalaan data
    if name == 'NeuralNetwork':
        scaler = StandardScaler()
        X_train_scaled = scaler.fit_transform(X_train)
        model.fit(X_train_scaled, Y_train)
    else:
        model.fit(X_train, Y_train)

    cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)
    results.append(cv_results)
    names.append(name)
    msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
    print(msg)

DecisionTree: 0.950000 (0.076376)
RandomForest: 0.966667 (0.040825)
NeuralNetwork: 0.983333 (0.033333)



Dalam proses pembangunan model klasifikasi, dimana melakukan eksplorasi untuk tiga algoritma berbeda dari contoh yang ada dimodul yaitu : Decision Tree, Random Forest, dan Neural Network. Evaluasi kinerja model dilakukan menggunakan K-Fold Cross Validation dengan 10 fold, di mana setiap model diuji sebanyak 10 kali dengan kombinasi data pelatihan dan validasi yang berbeda. Untuk Neural Network, dilakukan penskalaan data menggunakan StandardScaler. Hasil evaluasi menunjukkan bahwa Neural Network mencapai akurasi tertinggi sebesar 98.33%, dengan deviasi standar sekitar 3, lalu Decision Tree (95%) dan Random Forest (97.5%).Dari hasil output di atas, dapat dilihat bahwa Neural Network memiliki nilai akurasi perkiraan terbesar. Setelah mengetahui model yang paling akurat yaitu Neural Network, selanjutnya dapat dilakukan  pengujian tentang keakuratan model Neural Network terhadap data yang ada.

**IV. Memilih Model Terbaik**

In [20]:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [22]:
# Inisialisasi dan latih model Neural Network
neural_network_model = MLPClassifier(max_iter=1000, random_state=seed)
neural_network_model.fit(X_train, Y_train)

# prediksi pada dataset validasi
predictions_nn = neural_network_model.predict(X_validation)

# Evaluasi hasil prediksi
accuracy_nn = accuracy_score(Y_validation, predictions_nn)
confusion_matrix_nn = confusion_matrix(Y_validation, predictions_nn)
classification_report_nn = classification_report(Y_validation, predictions_nn)

# hasil evaluasi
print("Accuracy Neural Network:", accuracy_nn)
print("\nConfusion Matrix Neural Network:\n", confusion_matrix_nn)
print("\nClassification Report Neural Network:\n", classification_report_nn)

Accuracy Neural Network: 0.9666666666666667

Confusion Matrix Neural Network:
 [[ 7  0  0]
 [ 0 11  1]
 [ 0  0 11]]

Classification Report Neural Network:
                  precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       1.00      0.92      0.96        12
 Iris-virginica       0.92      1.00      0.96        11

       accuracy                           0.97        30
      macro avg       0.97      0.97      0.97        30
   weighted avg       0.97      0.97      0.97        30



Pada tahap evaluasi model Neural Network menggunakan dataset validasi, hasil menunjukkan akurasi sebesar 96.67%. Selanjutnya, Confusion matrix menggambarkan jumlah prediksi yang benar dan salah untuk setiap kelas, menunjukkan bahwa model menghasilkan prediksi yang baik untuk kelas Iris-setosa, Iris-versicolor, dan Iris-virginica, terutama untuk untuk kelas Iris-setosa. Precision, recall, dan f1-score pada classification report memberikan informasi tentang kinerja model untuk setiap kelas. Model Neural Network memiliki nilai precision, recall, dan f1-score yang tinggi untuk setiap kelas, menunjukkan kemampuan model dalam mengidentifikasi dan mengklasifikasikan dengan akurat.

**2. Buatlah evaluasi algoritma dengan dataset yang telah dicoba pada tugas
praktikum sebelumnya (dengan menggunakan 3 model yaitu KNN, NB dan
SVM)!**

In [23]:
import pandas as pd

In [24]:
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/glass/glass.data"
column_names = ['ID', 'RI', 'Na', 'Mg', 'Al', 'Si', 'K', 'Ca', 'Ba', 'Fe', 'Class']
dataset = pd.read_csv(url, names=column_names)

In [25]:
dataset.head(5)

ID       RI     Na    Mg    Al     Si     K    Ca   Ba   Fe  Class
0   1  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.0  0.0      1
1   2  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.0  0.0      1
2   3  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.0  0.0      1
3   4  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.0  0.0      1
4   5  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.0  0.0      1

In [26]:
dataset.shape

(214, 11)

**I. Validasi Dataset**

In [27]:
from sklearn import model_selection

In [28]:
# Split-out validation dataset
array = dataset.values

X = array[:,0:10]
Y = array[:,10]

validation_size = 0.20

seed = 7

X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

In [29]:
print("X_train:", X_train.shape)
print("X_validation:", X_validation.shape)
print("Y_train:", Y_train.shape)
print("Y_validation:", Y_validation.shape)

X_train: (171, 10)
X_validation: (43, 10)
Y_train: (171,)
Y_validation: (43,)


Pada tahap validasi dataset, dilakukan pembagian dataset menjadi dua subset: data pelatihan (training) dan data validasi. Dimana proses ini bertujuan untuk mengukur kinerja model pada data yang tidak digunakan selama pelatihan. Dapat dilihat diatas bahwa dataset awal dipecah menjadi X_train dan X_validation untuk fitur-fitur (atribut) serta Y_train dan Y_validation untuk label kelas. Setelah itu persentase atau bagian dari total validasi dataset diatur menggunakan parameter test_size yang disetel sebesar 20%, dengan nilai seed (random state) sebesar 7 untuk memastikan hasil. Hasil dari pembagian dataset ini terdapat 171 sampel dalam set pelatihan dan 43 sampel dalam set validasi.

**II. K-Folds Cross Validation**

In [30]:
# Test options and evaluation metric
seed = 7
scoring = 'accuracy'

K-Folds Cross Validation adalah teknik pemodelan data yang membagi data menjadi beberapa fold, kemudian melatih model pada fold tertentu dan mengevaluasinya pada fold lainnya. Hal ini dilakukan berulang kali hingga semua fold telah digunakan untuk pelatihan dan evaluasi.Parameter **seed** digunakan untuk memastikan bahwa hasil cross validation tetap konsisten setiap kali dijalankan dan untuk parameter **scoring** digunakan untuk menentukan metrik evaluasi yang akan digunakan untuk mengukur performa model. Dalam kasus ini, metrik yang digunakan adalah akurasi. Dengan menggunakan K-Folds Cross Validation yang nantinya akan mendapatkan lebih akurat tentang performa model dibandingkan dengan hanya menggunakan data latih dan data uji.

**III.Membangun Model**

In [31]:
# Import Library berisi algoritma klasifikasi yang akan digunakan
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC

# Spot Check Algorithms
models = []
models.append(('KNN', KNeighborsClassifier()))
models.append(('NB', GaussianNB()))
models.append(('SVM', SVC()))


In [32]:
# evaluate each model in turn
results = []

names = []

for name, model in models:
  kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)
  cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)
  results.append(cv_results)
  names.append(name)
  msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
  print(msg)

KNN: 0.964706 (0.047059)
NB: 0.877451 (0.054630)
SVM: 0.800654 (0.118704)


pada tahap ini dilakukan pembangunan model klasifikasi menggunakan 3 algortima yaitu  K-Nearest Neighbors (KNN), Naive Bayes (NB), dan Support Vector Machine (SVM). Evaluasi kinerja model dilakukan menggunakan K-Fold Cross Validation dengan 10 fold, di mana setiap model diuji sebanyak 10 kali dengan kombinasi data pelatihan dan validasi yang berbeda.Hasil evaluasi menunjukan bahwa model KNN memiliki akurasi tertinggi dengan rata-rata sebesar 96.47%, lalu Naive Bayes dengan akurasi rata-rata sebesar 87.74%, dan SVM dengan akurasi rata-rata sebesar 80.07%. Setelah mengetahui model yang paling akurat yaitu KNN, selanjutnya dapat dilakukan  pengujian tentang keakuratan model KNN  terhadap data yang ada.

**IV. Memilih Model Terbaik**

In [33]:
#import
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

In [34]:
knn_model = KNeighborsClassifier()

knn_model.fit(X_train, Y_train)

predictions = knn_model.predict(X_validation)

print(accuracy_score(Y_validation, predictions))
print(confusion_matrix(Y_validation, predictions))
print(classification_report(Y_validation, predictions))

1.0
[[10  0  0  0  0  0]
 [ 0 22  0  0  0  0]
 [ 0  0  3  0  0  0]
 [ 0  0  0  2  0  0]
 [ 0  0  0  0  1  0]
 [ 0  0  0  0  0  5]]
              precision    recall  f1-score   support

         1.0       1.00      1.00      1.00        10
         2.0       1.00      1.00      1.00        22
         3.0       1.00      1.00      1.00         3
         5.0       1.00      1.00      1.00         2
         6.0       1.00      1.00      1.00         1
         7.0       1.00      1.00      1.00         5

    accuracy                           1.00        43
   macro avg       1.00      1.00      1.00        43
weighted avg       1.00      1.00      1.00        43



Pada tahap evaluasi model KNN menggunakan dataset validasi, hasil menunjukkan akurasi sebesar 96.47%. Selanjutnya, Confusion matrix menggambarkan jumlah prediksi yang benar dan salah untuk setiap kelas, menunjukkan bahwa model menghasilkan prediksi yang baik untuk kelas pada dataset glass identifaction. Precision, recall, dan f1-score pada classification report memberikan informasi tentang kinerja model untuk setiap kelas.  Kesimpulannya, model KNN memberikan hasil yang baik pada dataset validasi, dengan akurasi 100%.